arbor/threading/cthread.cpp

```diff
@@ -11,132 +11,121 @@
 using namespace arb::threading::impl;
 using namespace arb;
 
-// RAII owner for a task in flight
-struct task_pool::run_task {
-    task_pool& pool;
-    lock& lck;
-    task tsk;
+bool notification_queue::try_pop(task& tsk) {
+    lock q_lock{q_mutex_, std::try_to_lock};
+    if (!q_lock || q_tasks_.empty()) return false;
+    tsk = std::move(q_tasks_.front());
+    q_tasks_.pop_front();
+    return true;
+}
```

Comment:
I think will make the code a bit uglier, for example in run_tasks_loop()

```
if(!tsk && !q_[i].pop(tsk)) break;
```

will become 

``` 
if(!tsk) {
    tsk = q_[i].pop();
    if (!tsk)
        break;
}
```

Reply:
It would look a bit different, but not quite as ugly as that I think :) e.g. (I haven;t tested this):
```
void task_system::run_tasks_loop(){
    size_t i = get_current_thread();
    while (true) {
        task tsk;
        for (unsigned n = 0; !tsk && n != count_; n++) {
            tsk = q_[(i + n) % count_].try_pop();
        }
        if (!tsk) tsk = q_[i].pop(tsk);
        if (!tsk) break;

        tsk();
    }
}
```